### PR TITLE
Extract generic database framework

### PR DESCRIPTION
**Pull request by aerokite**

### CODE REVIEW

File: pkg/controller/deleted_database.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+	"github.com/k8sdb/apimachinery/pkg/eventer"
+	kapi "k8s.io/kubernetes/pkg/api"
+	k8serr "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+)
+
+type Deleter interface {
```
Comment (TamalSaha) [inline pkg/controller/deleted_database.go]:
```
type Deleter interface {
	// Check Database TPR
	Exists(*tapi.DeletedDatabase) (bool, error)
	// Delete operation
	DeleteDatabase(*tapi.DeletedDatabase) error
	// Destroy operation
	DestroyDatabase(*tapi.DeletedDatabase) error
}
```

File: pkg/controller/database_snapshot.go [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type Snapshotter interface {
+	Validate(*tapi.DatabaseSnapshot) error
+	GetDatabaseRuntimeObject(*tapi.DatabaseSnapshot) (runtime.Object, error)
+	GetSnapshotObjects(*tapi.DatabaseSnapshot) (*kbatch.Job, error)
+	Destroy(*tapi.DatabaseSnapshot) error
```
Comment (TamalSaha) [inline pkg/controller/database_snapshot.go]: `DestroySnapshot(*tapi.DatabaseSnapshot) error`

File: pkg/controller/database_snapshot.go [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type Snapshotter interface {
+	Validate(*tapi.DatabaseSnapshot) error
+	GetDatabaseRuntimeObject(*tapi.DatabaseSnapshot) (runtime.Object, error)
```
Comment (TamalSaha) [inline pkg/controller/database_snapshot.go]: Why do you need this method? You should be able to get it from TPR directly.

Comment (aerokite) [inline pkg/controller/database_snapshot.go]: We need this to set events. Suppose you are taking snapshot for Elastic. You want to add events in that Elastic. So I need this runtime object.

File: pkg/controller/database_snapshot.go [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+	c.eventRecorder.PushEvent(
+		kapi.EventTypeNormal, eventer.EventReasonSuccessfulSnapshot, "Successfully completed snapshot",
+		runtimeObj, dbSnapshot,
+	)
+
+	go c.checkDatabaseSnapshotJob(dbSnapshot, job.Name, durationCheckSnapshotJob)
+}
+
+func (c *DatabaseSnapshotController) delete(dbSnapshot *tapi.DatabaseSnapshot) {
+	runtimeObj, _ := c.snapshoter.GetDatabaseRuntimeObject(dbSnapshot)
```
Comment (TamalSaha) [inline pkg/controller/database_snapshot.go]: Why ignore this error?

File: pkg/controller/database_snapshot.go [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type Snapshotter interface {
+	Validate(*tapi.DatabaseSnapshot) error
+	GetDatabaseRuntimeObject(*tapi.DatabaseSnapshot) (runtime.Object, error)
+	GetSnapshotObjects(*tapi.DatabaseSnapshot) (*kbatch.Job, error)
```
Comment (TamalSaha) [inline pkg/controller/database_snapshot.go]: GetSnapshotObjects -> TakeSnapshot

File: pkg/controller/database_snapshot.go (hunk 3 above)
Comment (TamalSaha) [inline pkg/controller/database_snapshot.go]: GetDatabaseRuntimeObject - > GetDatabase

**Pull request merged**